File: deno/src/lib/docker.ts
// Thin wrappers around `docker` CLI. We deliberately stay close to the
// shell-out shape — the bash version's contract is "we drive the docker
// CLI", and the deno port keeps that for compatibility with existing
// volumes/labels.

export type RunOptions = {
    args: string[];
    stdin?: "inherit" | "null";
    stdout?: "inherit" | "piped" | "null";
    stderr?: "inherit" | "piped" | "null";
};

// Run `docker <args>`, return its exit code, optionally captured stdout/err.
export async function dockerRun(opts: RunOptions): Promise<{
    code: number;
    stdout: string;
    stderr: string;
}> {
    const cmd = new Deno.Command("docker", {
        args: opts.args,
        stdin: opts.stdin ?? "inherit",
        stdout: opts.stdout ?? "inherit",
        stderr: opts.stderr ?? "inherit",
    });
    const { code, stdout, stderr } = await cmd.output();
    const dec = new TextDecoder();
    return {
        code,
        stdout: opts.stdout === "piped" ? dec.decode(stdout) : "",
        stderr: opts.stderr === "piped" ? dec.decode(stderr) : "",
    };
}

// Run docker, capture stdout. Throws on non-zero exit.
export async function dockerCapture(args: string[]): Promise<string> {
    const r = await dockerRun({ args, stdout: "piped", stderr: "piped" });
    if (r.code !== 0) {
        throw new Error(`docker ${args.join(" ")} failed (${r.code}): ${r.stderr.trim()}`);
    }
    return r.stdout;
}

// Inherit stdio for the foreground container — this is the long-running
// `docker run -it` invocation. Returns the container's exit code so the
// wrapper can propagate it.
export async function dockerExec(args: string[]): Promise<number> {
    const cmd = new Deno.Command("docker", {
        args,
        stdin: "inherit",
        stdout: "inherit",
        stderr: "inherit",
    });
    const { code } = await cmd.output();
    return code;
}

// docker volume create — idempotent, --label arguments accepted as labels.
// Returns the volume name on success.
export async function ensureVolume(name: string, labels: readonly string[] = []): Promise<void> {
    const args = ["volume", "create"];
    for (const l of labels) args.push("--label", l);
    args.push(name);
    await dockerCapture(args);
}

// One-shot mkdir inside a throwaway container so volume-subpath mounts have
// the expected sub-directory structure on first use.
export async function ensureSubpaths(
    image: string,
    volume: string,
    subpaths: readonly string[],
): Promise<void> {
    await dockerCapture([
        "run",
        "--rm",
        "--entrypoint",
        "mkdir",
        "-v",
        `${volume}:/v`,
        image,
        "-p",
        ...subpaths.map((s) => `/v/${s}`),
    ]);
}

File: src/docker.rs
use anyhow::{bail, Context, Result};
use std::process::{Command, Stdio};

// Capture stdout from a `docker <args>` invocation. Errors include captured
// stderr so the user sees the actual docker failure.
pub fn capture(args: &[&str]) -> Result<String> {
    let output = Command::new("docker")
        .args(args)
        .stdin(Stdio::null())
        .output()
        .context("spawning docker")?;
    if !output.status.success() {
        bail!(
            "docker {} failed ({}): {}",
            args.join(" "),
            output.status,
            String::from_utf8_lossy(&output.stderr).trim()
        );
    }
    Ok(String::from_utf8_lossy(&output.stdout).into_owned())
}

// Inherit stdio for the foreground container — this is the long-running
// `docker run -it` invocation. Returns the container's exit code so the
// wrapper can propagate it.
pub fn run_inherit(args: &[String]) -> Result<i32> {
    let status = Command::new("docker")
        .args(args)
        .status()
        .context("spawning docker")?;
    Ok(status.code().unwrap_or(1))
}

// `docker volume create` — idempotent; takes label strings as `key=value`.
pub fn ensure_volume(name: &str, labels: &[String]) -> Result<()> {
    let mut args: Vec<&str> = vec!["volume", "create"];
    for l in labels {
        args.push("--label");
        args.push(l);
    }
    args.push(name);
    capture(&args)?;
    Ok(())
}

// One-shot mkdir inside a throwaway container so volume-subpath mounts have
// the expected sub-directory structure on first use.
pub fn ensure_subpaths(image: &str, volume: &str, subpaths: &[&str]) -> Result<()> {
    let mut args: Vec<String> = ["run", "--rm", "--entrypoint", "mkdir", "-v"]
        .iter()
        .map(|s| s.to_string())
        .collect();
    args.push(format!("{volume}:/v"));
    args.push(image.to_string());
    args.push("-p".to_string());
    for s in subpaths {
        args.push(format!("/v/{s}"));
    }
    let refs: Vec<&str> = args.iter().map(|s| s.as_str()).collect();
    capture(&refs)?;
    Ok(())
}
